src/server/transaction.cc

```diff
@@ -1200,7 +1200,12 @@ size_t Transaction::ReverseArgIndex(ShardId shard_id, size_t arg_index) const {
 
 OpStatus Transaction::WaitOnWatch(const time_point& tp, WaitKeysProvider wkeys_provider,
                                   KeyReadyChecker krc, bool* block_flag, bool* pause_flag) {
-  DCHECK(!blocking_barrier_.IsClaimed());  // Blocking barrier can't be re-used
+  if (blocking_barrier_.IsClaimed()) {  // Might have been cancelled ahead by a dropping connection
+    Conclude();
+    return OpStatus::CANCELLED;
+  }
```

Comment:
Also, just removing the DCHECK is an option.

This would have happened in the general flow either way because Wait() below would immediately, it's just that we do some more operations for registering and unregistering immediately the blocking stuff. Not sure what option is better 🤷🏻‍♂️